codice/ddf
# DDF-5802 Solr Split Catalog Provider

#### What does this PR do?
address improvement in
https://github.com/codice/ddf/issues/5802

The SplitCatalogProvider implements the CatalogProvider interface. It manages the configuration and connection to multiple solr collections.  SolrCollectionProvider implementations are chained together to process each modify request. During ingest/update the metacard tag is parsed by the SolrCollectionProvider to determine which solr collection to create/update. During query, the solr alias is used to query across all collections. Special logic is added to handle realtime get query which doesn't with alias

1. Rename SolrCatalogProvider to SolrCatalogProviderImpl
2. Adding timeout for isAvailable check
3. SolrCloudClientFactory handling of multiple collection configs
4. Add SplitCatalogProvider to implement CatalogProvider interface
5. Implement a DefaultSolrCollectionProvider as the default SolrCollectionProvider implementation
6. Add split profile 
7. Add TestSplitCatalog itest which extends TestCatalog

Who is reviewing it?

Select relevant component teams:
@codice/core-apis
@codice/solr

Ask 2 committers to review/merge the PR and tag them here.
@pklinef
@bdeining
@millerw8
@paouelle 
@leo-sakh 

#### How should this be tested?


- Start up solrCloud docker
-  profile:Install split
- Verify catalog_default collection is created
- Ingest some record and query
- Verify some record is in catalog_default collection
- http://localhost:8994/solr/catalog_default/select?q=*%3A*
- http://localhost:8994/solr/catalog/select?q=*%3A*
- Verify alias is associated with catalog_default collection
- http://localhost:8994/solr/admin/collections?action=LISTALIASES&wt=xml

#### Any background context you want to provide?
This is part 1/3 of this effort
https://connexta.atlassian.net/wiki/spaces/ARCH/pages/758939649/Ingest+Query+Performance+Enhancements

1. Next issue https://github.com/codice/ddf/issues/5859  will add a SuggesterCollectionProvider, this will provides a new CollectionProvider to handle all gazetteer data on a separate solr collection
2. Implement NtrCollectionProvider to handler all user related data https://github.com/codice/ddf/issues/5860

#### What are the relevant tickets?
Fixes: #____

#### Screenshots


#### Checklist:
- [ x ] Update / Add Unit Tests
- [ x ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
